File: packages/depz-sensor-sdk-ts/test/fake-vl53l8.ts
/**
 * Fake VL53L8 register/DCI platform for host-side ULD tests (no hardware).
 * Port of the Python `tests/fake_vl53l8.py`.
 *
 * Emulates the register bridge + DCI transport (swapBuffer + header/footer
 * framing) so the real byte sequences in uld.ts are exercised end to end.
 */

import {
  FW_CHECKSUM,
  OFFSET_BUFFER_SIZE,
  VL53L8CX,
  XTALK_BUFFER_SIZE,
  swapBuffer,
  type Vl53l8Platform,
} from "../src/sensors/vl53l8/uld.js";

const UI_CMD_STATUS = 0x2c00;
const UI_CMD_START = 0x2c04;
const UI_CMD_END = 0x2fff;

export class FakeUldPlatform implements Vl53l8Platform {
  readonly reg = new Map<number, number>();
  readonly dci = new Map<number, Uint8Array>();
  /** Every wrMulti (addr, bytes) — for transcript assertions. */
  readonly writes: Array<[number, Uint8Array]> = [];
  private pendingRead: [number, number] | null = null;

  sleepMs(): Promise<void> {
    return Promise.resolve();
  }

  wrMulti(addr: number, data: Uint8Array): Promise<void> {
    const buf = data.slice();
    this.writes.push([addr, buf]);
    // DCI read command: 12 bytes written to UI_CMD_END-11 (0x2FF4).
    if (addr === UI_CMD_END - 11 && buf.length === 12) {
      const index = (buf[0]! << 8) | buf[1]!;
      const size = ((buf[2]! & 0xff) << 4) | ((buf[3]! & 0xff) >> 4);
      this.pendingRead = [index, size];
      return Promise.resolve();
    }
    // DCI write: header(4) + swap(payload) + footer(8); footer sig 0F 05 01.
    const n = buf.length;
    if (n >= 12 && buf[n - 5] === 0x0f && buf[n - 4] === 0x05 && buf[n - 3] === 0x01) {
      const index = (buf[0]! << 8) | buf[1]!;
      const dataSize = ((buf[2]! & 0xff) << 4) | ((buf[3]! & 0xff) >> 4);
      this.dci.set(index, swapBuffer(buf.subarray(4, 4 + dataSize)));
      return Promise.resolve();
    }
    // plain byte writes → registers
    for (let i = 0; i < n; i++) this.reg.set(addr + i, buf[i]!);
    return Promise.resolve();
  }

  rdMulti(addr: number, size: number): Promise<Uint8Array> {
    if (addr === UI_CMD_STATUS) {
      const out = new Uint8Array(size);
      const base = [0x00, 0x03, 0x00, 0x00];
      for (let i = 0; i < Math.min(size, 4); i++) out[i] = base[i]!;
      return Promise.resolve(out);
    }
    if (addr === UI_CMD_START && this.pendingRead !== null) {
      const [index, dsize] = this.pendingRead;
      const stored = this.dci.get(index) ?? new Uint8Array(dsize);
      const s = new Uint8Array(dsize);
      s.set(stored.subarray(0, dsize));
      const raw = new Uint8Array(dsize + 12);
      raw.set(s, 4);
      return Promise.resolve(swapBuffer(raw));
    }
    const out = new Uint8Array(size);
    for (let i = 0; i < size; i++) out[i] = this.reg.get(addr + i) ?? 0;
    return Promise.resolve(out);
  }
}

/** Construct a VL53L8CX bypassing blob-file init (mirrors Python make_driver). */
export function makeDriver(): { drv: VL53L8CX; p: FakeUldPlatform } {
  const p = new FakeUldPlatform();
  const drv = Object.create(VL53L8CX.prototype) as VL53L8CX;
  const a = drv as unknown as Record<string, unknown>;
  a.p = p;
  a.variant = "cx";
  a.fwChecksum = FW_CHECKSUM.cx;
  a.outputEnableW3 = 0xc0000000;
  a.frameTail = 32;
  a.footerIdOff = 12;
  a.motionPresent = false;
  a.defaultXtalk = new Uint8Array(XTALK_BUFFER_SIZE);
  a.offsetData = new Uint8Array(OFFSET_BUFFER_SIZE);
  a.xtalkData = new Uint8Array(XTALK_BUFFER_SIZE);
  a.streamcount = 255;
  a.dataReadSize = 0;
  a.lastBlocks = [];
  return { drv, p };
}

File: packages/depz-sensor-sdk-python/tests/fake_vl53l8.py
"""Fake VL53L8 register/DCI platform for host-side ULD tests (no hardware).

Emulates the register bridge + DCI transport (swap_buffer + header/footer
framing) so the real byte sequences in uld.py are exercised end to end.
"""

from __future__ import annotations

from depz_sensor_sdk.vl53l8 import uld
from depz_sensor_sdk.vl53l8.uld import VL53L8CX, swap_buffer

UI_CMD_STATUS = 0x2C00
UI_CMD_START = 0x2C04
UI_CMD_END = 0x2FFF


class FakeUldPlatform:
    """Emulates the register bridge + DCI transport for uld.VL53L8CX."""

    def __init__(self):
        self.reg: dict[int, int] = {}
        self.dci: dict[int, bytes] = {}
        self._pending_read: tuple[int, int] | None = None
        self.writes: list[tuple[int, bytes]] = []

    def sleep_ms(self, ms):
        pass

    def wr_multi(self, addr: int, data: bytes) -> None:
        data = bytes(data)
        self.writes.append((addr, data))
        # DCI read command: 12 bytes written to UI_CMD_END-11 (0x2FF4)
        if addr == UI_CMD_END - 11 and len(data) == 12:
            index = (data[0] << 8) | data[1]
            size = ((data[2] & 0xFF) << 4) | ((data[3] & 0xFF) >> 4)
            self._pending_read = (index, size)
            return
        # DCI write: header(4) + swap(payload) + footer(8), footer signature
        if len(data) >= 12 and data[-5] == 0x0F and data[-4] == 0x05 and data[-3] == 0x01:
            index = (data[0] << 8) | data[1]
            data_size = ((data[2] & 0xFF) << 4) | ((data[3] & 0xFF) >> 4)
            payload = swap_buffer(data[4 : 4 + data_size])
            self.dci[index] = payload
            return
        # plain byte writes → registers
        for i, b in enumerate(data):
            self.reg[addr + i] = b

    def rd_multi(self, addr: int, size: int) -> bytes:
        if addr == UI_CMD_STATUS:
            return bytes([0x00, 0x03, 0x00, 0x00])[:size].ljust(size, b"\x00")
        if addr == UI_CMD_START and self._pending_read is not None:
            index, dsize = self._pending_read
            stored = self.dci.get(index, b"\x00" * dsize)
            stored = stored[:dsize].ljust(dsize, b"\x00")
            raw = (b"\x00" * 4 + stored).ljust(dsize + 12, b"\x00")
            return swap_buffer(raw)
        return bytes([self.reg.get(addr + i, 0) for i in range(size)])


def make_driver():
    """Return (driver, platform) with a VL53L8CX bypassing blob-file init."""
    p = FakeUldPlatform()
    drv = VL53L8CX.__new__(VL53L8CX)
    drv.p = p
    drv.variant = "cx"
    drv._output_enable_w3 = 0xC0000000
    drv._frame_tail = 32
    drv._footer_id_off = 12
    drv._motion_present = False
    drv.default_xtalk = b"\x00" * uld.XTALK_BUFFER_SIZE
    return drv, p
